### PR TITLE
add warning when kubectl set env from secret leads to uppercase

### LINKED ISSUE #1171: Show warning when kubectl set env from secret leads to uppercase of the secret keys
### What happened?

I have a **Secret** created by the following command: 
```
kubectl -n <namespace-name> create secret generic xsuaa-credentials \
--from-literal=url='<xsuaa-url>' \
--from-literal=clientid='<xsuaa-clientid>' \
--from-literal=clientsecret='<xsuaa-clientsecret>' \
--from-literal=verificationkey='<xsuaa-verificationkey>' \
--from-literal=xsappname='<xsuaa-xsappname>'
```
> **_Note:_** The secret keys are in lowercase.

I'm binding the secret to a **Deployment** with the following command:

```
kubectl -n <namespace-name> set env --from=secret/xsuaa-credentials deployment/xsk
```

When describing the deployment I get the following output:

```
Name:                   xsk
Namespace:              xsk-demo-sdi
CreationTimestamp:      Fri, 14 Jan 2022 18:27:12 +0200
Labels:                 <none>
Annotations:            deployment.kubernetes.io/revision: 2
                        kubectl.kubernetes.io/last-applied-configuration:
                          {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"name":"xsk","namespace":"xsk-demo-sdi"},"spec":{"replicas":1,"se...
Selector:               app=xsk
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:  app=xsk
  Containers:
   xsk:
    Image:      dirigiblelabs/xsk-kyma:latest
    Port:       8080/TCP
    Host Port:  0/TCP
    Environment:
      URL:                      <set to the key 'url' in secret 'xsuaa-credentials'>              Optional: false
      VERIFICATIONKEY:          <set to the key 'verificationkey' in secret 'xsuaa-credentials'>  Optional: false
      XSAPPNAME:                <set to the key 'xsappname' in secret 'xsuaa-credentials'>        Optional: false
      CLIENTID:                 <set to the key 'clientid' in secret 'xsuaa-credentials'>         Optional: false
      CLIENTSECRET:             <set to the key 'clientsecret' in secret 'xsuaa-credentials'>     Optional: false
```

Where the environment variables are in uppercase:
- clientid &#8594; CLIENTID
- clientsecret &#8594; CLIENTSECRET
- verificationkey &#8594; VERIFICATIONKEY
- xsappname &#8594; XSAPPNAME

### What did you expect to happen?

The environment variables to be in lowercase, as they are defined in the secret.

### How can we reproduce it (as minimally and precisely as possible)?

Shown in the description.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-19T22:12:47Z", GoVersion:"go1.12.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.6", GitCommit:"d921bc6d1810da51177fbd0ed61dc811c5228097", GitTreeState:"clean", BuildDate:"2021-10-27T17:44:26Z", GoVersion:"go1.16.9", Compiler:"gc", Platform:"linux/amd64"}

```

</details>

### Cloud provider

<details>
AWS via Gardener
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by jlsong01**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Show warning when kubectl set env from secret leads to uppercase of the secret keys.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1171

#### Special notes for your reviewer:
some go format refinement included.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by jlsong01** [comment]
@seans3 Hey Sean, could you please take a look? Thanks

File: staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go [hunk 1]
```diff
@@ -305,6 +305,12 @@ func (o *EnvOptions) RunEnv() error {
 			case *v1.Secret:
 				for key := range from.Data {
 					if contains(key, o.Keys) {
+						for _, l := range key {
```
Comment (seans3) [inline staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go:308]: You will probably want to pull this into it's own function.

**Review by seans3 — COMMENTED** [review]
The concept of adding a warning in `kubectl` when the environment variable is upper-cased is a good idea. As mentioned in the code, it is cleaner to pull this validation code into its own function. Importantly, you should add a unit test to prove the code works. Since the code outside `kubectl` that you've modified is only formatting changes, you may want to remove those changes so you won't have to get reviewers from other SIG's involved.

Thanks for taking the initiative on this.

File: staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go (hunk 1 above)
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go:308]: Agreed.  What about just using `keyToEnvName()`, and print a warning anytime keyToEnvName(key) != key?  I don’t think there is a need to iterate over each character.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/registry.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ type Resolver interface {
 func NewRegistry(stdExprEnv *cel.Env) *Registry {
 	return &Registry{
 		exprEnvs: map[string]*cel.Env{"": stdExprEnv},
-		schemas: map[string]*schema.Structural{},
```
Comment (brianpursley) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/registry.go:36]: Did you mean to change these other files?  They seem unrelated.

Comment (jlsong01) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/registry.go:36]: They are just some unrelated format refinement. As seans3 explained, I will revert them.

Comment (brianpursley) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/registry.go:36]: Sorry, didn’t see that part of his comment.

File: staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go (hunk 1 above)
Comment (jlsong01) [reply staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go]: a warning added into `keyToEnvName()`

Comment (jlsong01) [reply staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go]: unit cases have been added acoordingly

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/registry.go (hunk 2 above)
Comment (jlsong01) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/registry.go]: no problem and reverted :-)

**Review by seans3 — COMMENTED** [review]
Small nit about exporting, and then we can merge. Thanks for adding tests.

**Comment by jlsong01** [comment]
> Small nit about exporting, and then we can merge. Thanks for adding tests.

Sorry, I may not get your point. I assume that you mean change `KeyToEnvName` as `keyToEnvName`, to make it as a non-exported method. And I have updated it accordingly.

If my assumption is wrong, please let me know and I will rework it :-)

**Review by seans3 — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go [hunk 3]
```diff
@@ -207,6 +203,14 @@ func contains(key string, keyList []string) bool {
 	return false
 }
 
+func (o *EnvOptions) keyToEnvName(key string) string {
+	envName := strings.ToUpper(validEnvNameRegexp.ReplaceAllString(key, "_"))
+	if envName != key {
+		fmt.Fprintf(o.ErrOut, "warning: key %s transferred to %s\n", key, envName)
```
Comment (sftim) [inline staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go:209]: Do these end up rendered similarly to the way https://kubernetes.io/blog/2020/09/03/warnings/ explains? (colored yellow, and with an initial capital W)?

Comment (jlsong01) [reply staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go:209]: No, it's just normal font color, without colored yellow and an Initial capital W.

The warning style exists in several places. If you search with `fmt.Fprintf(o.ErrOut, "warning: xxxx`, you can find them.

Do we need coordinate them to the way as you mentioned? If so, I am glad to issue another PR to fix them all

Comment (sftim) [reply staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go:209]: > [cleanup]

Sounds good to me. How about first filing a kubectl issue to track the work? See https://github.com/kubernetes/kubectl/issues/new/choose

Comment (jlsong01) [reply staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go:209]: Sure, track issue first :-)

Comment (jlsong01) [reply staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go:209]: FYI - https://github.com/kubernetes/kubectl/issues/1177
